src/sdl2/render.rs

```diff
@@ -944,112 +999,49 @@ impl Renderer {
-                                    raw: old_texture_raw,
-                                    _marker: PhantomData,
-                                })
-                       }
-                   })
-            } else {
-                Err(get_error())
-            }
-        }
+pub type TextureRenderer<'r, 't> = Renderer<TextureTarget<'r, 't>>;
+
+impl<'r, 't> TextureRenderer<'r, 't> {
+    pub fn with_target<'nt, 'a>(mut self,
+                                texture: &'nt mut Texture<'a>)
+                                -> Result<TextureRenderer<'r, 'nt>, IntegerOrSdlError> {
+        unsafe { self.raw_target(texture.raw) }?;
+        let context = mem::replace(&mut self.context, unsafe { mem::zeroed() });
+        mem::forget(self);
+        Ok(Renderer {
+               context: context,
+               target: TextureTarget {
+                   _renderer_marker: PhantomData,
+                   _texture_marker: PhantomData,
+               },
+           })
     }
+}
 
-    /// Sets the render target to the provided texture.
-    /// The texture must be created with the texture access: `sdl2::render::TextureAccess::Target`.
-    ///
-    /// The old render target is returned if the function is successful.
-    pub fn set(&mut self, texture: Texture) -> Result<Option<Texture<'r>>, String> {
-        unsafe {
-            let old_texture_raw = ll::SDL_GetRenderTarget(self.raw);
-
-            if ll::SDL_SetRenderTarget(self.raw, texture.raw) == 0 {
-                Ok(match old_texture_raw.is_null() {
-                       true => None,
-                       false => {
-                           Some(Texture {
-                                    raw: old_texture_raw,
-                                    _marker: PhantomData,
-                                })
-                       }
-                   })
-            } else {
-                Err(get_error())
-            }
-        }
-    }
-
-    /// Creates a new texture and sets it as the render target.
-    ///
-    /// The old render target is returned if the function is successful.
-    pub fn create_and_set(&mut self,
-                          format: pixels::PixelFormatEnum,
-                          width: u32,
-                          height: u32)
-                          -> Result<Option<Texture<'r>>, IntegerOrSdlError> {
-        use common::IntegerOrSdlError::*;
-        let width = try!(validate_int(width, "width"));
-        let height = try!(validate_int(height, "height"));
-
-        let new_texture_raw = unsafe {
-            let access = ll::SDL_TEXTUREACCESS_TARGET;
-            ll::SDL_CreateTexture(self.raw, format as uint32_t, access as c_int, width, height)
-        };
-
-        if new_texture_raw == ptr::null_mut() {
-            Err(SdlError(get_error()))
-        } else {
+impl<T: Target> Drop for Renderer<T> {
+    // `Drop` cannot be specialized. Get around this through run-time check of Target Kind
```

Comment:
I briefly read the discussion in the specialization RFC and it looked like it was still an open question on whether Drop should allow specialization or not. It might be worth testing after it lands since it would clean up the code a bit (remove the need for the `Target` trait and the `TargetKind` enum).